comps/cores/mega/gateway.py

```diff
@@ -1,394 +1,411 @@
+            host=self.host,
+            port=self.port,
+            endpoint=self.endpoint,
+            input_datatype=self.input_datatype,
+            output_datatype=self.output_datatype,
+        )
+        self.define_routes()
+        self.service.start()
+
+    def define_routes(self):
+        self.service.app.router.add_api_route(self.endpoint, self.handle_request, methods=["POST"])
+        self.service.app.router.add_api_route(str(MegaServiceEndpoint.LIST_SERVICE), self.list_service, methods=["GET"])
+        self.service.app.router.add_api_route(
+            str(MegaServiceEndpoint.LIST_PARAMETERS), self.list_parameter, methods=["GET"]
+        )
+
+    def add_route(self, endpoint, handler, methods=["POST"]):
+        self.service.app.router.add_api_route(endpoint, handler, methods=methods)
+
+    def stop(self):
+        self.service.stop()
+
+    async def handle_request(self, request: Request):
+        raise NotImplementedError("Subclasses must implement this method")
+
+    def list_service(self):
+        response = {}
+        for node in self.all_leaves():
+            response = {self.services[node].description: self.services[node].endpoint_path}
+        return response
+
+    def list_parameter(self):
+        pass
+
+    def _handle_message(self, messages):
+        if isinstance(messages, str):
+            prompt = messages
+        else:
+            messages_dict = {}
+            for message in messages:
+                msg_role = message["role"]
+                if msg_role == "system":
+                    system_prompt = message["content"]
+                elif msg_role == "user":
+                    if type(message["content"]) == list:
+                        text = ""
+                        text_list = [item["text"] for item in message["content"] if item["type"] == "text"]
+                        text += "\n".join(text_list)
+                        messages_dict[msg_role] = text
+                    else:
+                        messages_dict[msg_role] = message["content"]
+                elif msg_role == "assistant":
+                    messages_dict[msg_role] = message["content"]
+                else:
+                    raise ValueError(f"Unknown role: {msg_role}")
+            prompt = system_prompt + "\n"
+            for role, message in messages_dict.items():
+                if message:
+                    prompt += role + ": " + message + "\n"
+                else:
+                    prompt += role + ":"
+        return prompt
+
+
+class ChatQnAGateway(Gateway):
+    def __init__(self, megaservice, host="0.0.0.0", port=8888):
+        super().__init__(
+            megaservice, host, port, str(MegaServiceEndpoint.CHAT_QNA), ChatCompletionRequest, ChatCompletionResponse
+        )
+
+    async def handle_request(self, request: Request):
+        data = await request.json()
+        stream_opt = data.get("stream", True)
+        chat_request = ChatCompletionRequest.parse_obj(data)

```

Comment:
The `ChatCompletionRequest` doesn't define retriever and reranker parameters. How can we customize them?